### PR TITLE
Product Review - Week 2

### PR DESCRIPTION
**Pull request by yeonjupark95**

### CODE REVIEW

File: api/orders.js [hunk 1]
```diff
@@ -1,4 +1,232 @@
...
+      });
+    }
+  }
+);
+
+// Create a new order on visit and when there is no existing cart. Create a new order when the order status changes to success.
+// Order default status is order_pending.
+// endpoint "/cart"
+ordersRouter.post("/", async (req, res, next) => {
+  const { email, address } = req.body;
```
Comment (rpuzon01) [inline api/orders.js:101]: what if there was no email or address

File: api/orders.js [hunk 2]
```diff
@@ -1,4 +1,232 @@
...
+      res.send(newOrder);
+    } else if (req.user) {
+      const { id } = req.user;
+      const newOrder = await createOrder({
+        userId: id,
+        email,
+        address,
+      });
+      res.send(newOrder);
+    }
```
Comment (rpuzon01) [inline api/orders.js:118]: i'd like to see comments describing what is happening in the if and else blocks

File: api/orders.js [hunk 3]
```diff
@@ -1,4 +1,232 @@
...
+    const orderStatus = await setOrderAsPaymentPending(id);
+    res.send(orderStatus);
+  } catch (error) {
+    console.error(error);
+    next({
+      name: "OrderStatusProcessingError",
+      message: "Failed to update the order as processing",
+    });
+  }
+});
```
Comment (rpuzon01) [inline api/orders.js:141]: not needed with stripe

File: api/orders.js [hunk 4]
```diff
@@ -1,4 +1,232 @@
...
+      next({
+        name: "OrderStatusSuccessError",
+        message: "Failed to update the order as success",
+      });
+    }
+  }
+);
+
+// This is to update any order info such as email and address. This is unrelated to status updates.
+ordersRouter.patch("/update/:orderId", async (req, res, next) => {
```
Comment (rpuzon01) [inline api/orders.js:179]: doesn't follow RESTful api standard

```suggestion
ordersRouter.patch("/:orderId", async (req, res, next) => {
```

Comment (rpuzon01) [reply api/orders.js:179]: maybe add requireUser

File: api/products.js [hunk 5]
```diff
@@ -1,4 +1,73 @@
...
+    } catch (error) {
+        next(error)
+    }
+})
+
+productsRouter.get('/:name', async (req, res, next) => {
+    const { name }  = req.params;
+
+    try {
+        const product = await getProductByName(name);
```
Comment (rpuzon01) [inline api/products.js:19]: what if there was no product under that name

File: api/products.js [hunk 6]
```diff
@@ -1,4 +1,73 @@
...
+    const { name, description, category, quantity, price, photo } = req.body
+    try {
+        const addedProduct = await createProduct({ name, description, category, quantity, price, photo })
+        res.send(addedProduct)
+    } catch (error) {
+        next(error)
+    }
+})
+
+productsRouter.patch("/edit/:productId", requireUser, requireAdmin, async (req, res, next) => {
```
Comment (rpuzon01) [inline api/products.js:47]:
```suggestion
productsRouter.patch("/:productId", requireUser, requireAdmin, async (req, res, next) => {
```

File: api/products_orders.js [hunk 7]
```diff
@@ -1,4 +1,82 @@
...
+      name: "failedProductsOrdersUpdate",
+      message: "Failed to update products_orders",
+    });
+  }
+});
+
+products_ordersRouter.delete("/:products_orderId", async (req, res, next) => {
+  const { products_orderId } = req.params;
+  try {
+    const Products_Order = await deleteProductOrder(products_orderId);
```
Comment (rpuzon01) [inline api/products_orders.js:71]: never capitalize regular varaibles
```suggestion
    const products_Order = await deleteProductOrder(products_orderId);
```

File: api/reviews.js [hunk 8]
```diff
@@ -1,4 +1,117 @@
...
+  }
+});
+
+//User must be able to create a review for a particular product 
+reviewsRouter.post("/", requireUser, async (req, res, next) => {
+  const { userId, productId, description, rating } = req.body;
+  try {
+    const newReview = await createReview({
+      userId: userId,
+      productId: productId,
```
Comment (rpuzon01) [inline api/reviews.js:63]: you don't have use the : format if the name and the value are the same
```suggestion
      userId,
      productId,
```

File: api/reviews.js [hunk 9]
```diff
@@ -1,4 +1,117 @@
...
+    next(error);
+  }
+});
+
+//User must be able to make changes in a review when needed.
+reviewsRouter.patch("/:reviewId", requireUser, async (req, res, next) => {
+  const { reviewId } = req.params;
+  const { description, rating } = req.body;
+  try {
+    const reviewById = await getReviewById(reviewId);
```
Comment (rpuzon01) [inline api/reviews.js:79]: what if there's no review

File: api/users.js [hunk 10]
```diff
@@ -63,7 +73,8 @@ usersRouter.post("/login", async (req, res, next) => {
   }
 
   try {
-    const user = await getUser(req.body);
+    const user = await getUser({ username, password });
+    console.log("api user", user);
```
Comment (rpuzon01) [inline api/users.js:77]: bad

File: api/users.js [hunk 11]
```diff
@@ -63,7 +73,8 @@ usersRouter.post("/login", async (req, res, next) => {
   }
 
   try {
-    const user = await getUser(req.body);
+    const user = await getUser({ username, password });
```
Comment (rpuzon01) [inline api/users.js:76]: this is really good for white listing the body of the request.

File: api/users.js [hunk 12]
```diff
@@ -83,30 +94,35 @@ usersRouter.post("/login", async (req, res, next) => {
...
+  } catch (error) {
+    next(error);
+  }
 });
 
 //PATCH /users/me(*)
-usersRouter.patch("/me", requireUser, async (req, res, next) => {
+usersRouter.patch("/myaccount", requireUser, async (req, res, next) => {
+  const { id } = req.user;
+  const { ...userValuesToUpdate } = req.body;
```
Comment (rpuzon01) [inline api/users.js:110]: let me find a better way for this

File: api/users.js [hunk 13]
```diff
@@ -83,30 +94,35 @@ usersRouter.post("/login", async (req, res, next) => {
...
-    if (!authorization) {
-      return next({
-        name: "InvalidUserCannotUpdate",
+    const { id: userId } = await getUserById(id);
+    if (id !== userId) {
+      next({
+        name: "InvalidUserError",
         message: "You are not the owner of this account",
       });
     }
```
Comment (rpuzon01) [inline api/users.js:119]: you could prolly get rid of this

File: db/models/orders.js [hunk 14]
```diff
@@ -40,48 +41,33 @@ const getAllOrders = async () => {
...
-  }
-}
-
-const getOrdersByUser = async ({ username }) => {
+const getOrdersByUser = async (username) => {
   try {
     const { rows: orders } = await client.query(
       `
-            SELECT orders.*, users.username AS "customerName"
+            SELECT orders.*, users.username, users.id
```
Comment (rpuzon01) [inline db/models/orders.js:70]: the users id will take over the property of id
so if i have an object like this  
{
  id: userId => orderId
}

Comment (rpuzon01) [inline db/models/orders.js:70]:
```suggestion
            SELECT orders.*, users.username, users.id AS userId
```

File: db/models/orders.js [hunk 15]
```diff
@@ -154,15 +140,72 @@ const updateOrder = async ({ id, ...fields }) => {
...
+            WHERE id = $1
+            RETURNING *;
+        `,
+      [orderId]
+    );
+    return order;
+  } catch (error) {
+    throw error;
+  }
+};
```
Comment (rpuzon01) [inline db/models/orders.js:179]: won't need this anymore

File: db/models/products.js [hunk 16]
```diff
@@ -1,114 +1,148 @@
...
+};
 
 const getProductById = async (id) => {
-    try {
-        const { rows: [product] } = await client.query(`
+  try {
+    const {
+      rows: [product],
+    } = await client.query(
+      `
```
Comment (rpuzon01) [inline db/models/products.js:20]: indentation differences, Ryan TODO: look  for ways to ignore this

File: db/models/user.js [hunk 17]
```diff
@@ -1,10 +1,29 @@
...
+  }
+}
+
+const createUser = async ({ full_name, email, username, password }) => {
+  try {
+    if (!username || !password) {
+      throw {
+        name: "MissingFields",
+        message: "Please provide a username and a password",
+      };
```
Comment (rpuzon01) [inline db/models/user.js:22]: 😄

File: index.js [hunk 18]
```diff
@@ -1,41 +1,73 @@
...
+      mode: "payment",
+      success_url: `${process.env.SERVER_URL}/success.html`,
+      cancel_url: `${process.env.SERVER_URL}/cancel.html`,
+    });
+    console.log("session url:", { url: session.url });
+    res.send({ url: session.url });
+  } catch (error) {
+    console.error(error);
+  }
+});
```
Comment (rpuzon01) [inline index.js:70]: you should notttttttttttttttt have this here

File: index.js [hunk 19]
```diff
@@ -1,41 +1,73 @@
...
 
-// by default serve up the react app if we don't recognize the route
-server.use((req, res, next) => {
-  res.sendFile(path.join(__dirname, 'build', 'index.html'));
-});
+// bring in the DB connection
+const { client, getProductById, getOrderById } = require("./db");
+
+// here's our stripe
+const stripe = require("stripe")(process.env.STRIPE_PRIVATE_KEY);
```
Comment (rpuzon01) [inline index.js:29]: this also should not be here

File: public/cancel.html [hunk 20]
```diff
@@ -0,0 +1,12 @@
...
+<head>
+    <meta charset="UTF-8">
+    <meta http-equiv="X-UA-Compatible" content="IE=edge">
+    <meta name="viewport" content="width=device-width, initial-scale=1.0">
+    <title>Document</title>
+</head>
+<body>
+    
+</body>
+</html>
```
Comment (rpuzon01) [inline public/cancel.html:12]: this should be a react component

File: src/axios-services/index.js [hunk 21]
```diff
@@ -1,4 +1,6 @@
 import axios from 'axios';
+import { copyDone } from 'pg-protocol/dist/messages';
```
Comment (rpuzon01) [inline src/axios-services/index.js:2]: unneeded import

File: src/axios-services/index.js [hunk 22]
```diff
@@ -20,10 +22,118 @@ import axios from 'axios';
...
+    })
+    return data;
+  } catch (err) {
+    console.error("Error at deleteReview", err)
+  }
+}
+
+export const fetchOrder = async (id) => {
+  try {
+    const { data: order } = await axios.get(`${BASE_URL}/orders/:orderId`);
```
Comment (rpuzon01) [inline src/axios-services/index.js:102]:
```suggestion
    const { data: order } = await axios.get(`${BASE_URL}/orders/${id}`);
```

File: src/axios-services/index.js [hunk 23]
```diff
@@ -20,10 +22,118 @@ import axios from 'axios';
...
+    return order;
+  } catch (error) {
+    console.error(error);
+  }
+};
+
+export const createPendingOrder = async (email, address) => {
+  try {
+    const {
+      data: { pendingOrder },
```
Comment (rpuzon01) [inline src/axios-services/index.js:112]: double check if this works

File: src/components/Cart.js [hunk 24]
```diff
@@ -0,0 +1,35 @@
...
+        method: "POST",
+        headers: {
+          "Content-Type": "application/json",
+        },
+        body: JSON.stringify({
+          items: [
+            { id: 1, quantity: 3 },
+            { id: 2, quantity: 1 },
+          ],
+        }),
```
Comment (rpuzon01) [inline src/components/Cart.js:17]: this is hardcoded

File: src/components/Navigation.js [hunk 25]
```diff
@@ -0,0 +1,41 @@
...
+import { Link } from "react-router-dom";
+import "../style/Navigation.css";
+import account from "./img/account.png";
+import cart from "./img/cart.png";
+
+const Navigation = () => {
+  return (
+    <>
+      <div className="nav-container">
+        <Link to="/" style={{ textDecoration: "none" }}>
```
Comment (rpuzon01) [inline src/components/Navigation.js:11]: usually in-line styling is not recommended

File: src/components/Navigation.js [hunk 26]
```diff
@@ -0,0 +1,41 @@
...
+          <div className="plant-categories-link">Small Plants</div>
+          </Link>
+        </div>
+        <div className="account">
+          <a href="/myaccount">
+            <img className="nav-icon" src={account} alt="avatar-account-icon" />
+          </a>
+          <a href="/cart">
+            <img className="nav-icon" src={cart} alt="shopping-cart-icon" />
+          </a>
```
Comment (rpuzon01) [inline src/components/Navigation.js:34]: anchor tags will cause a refresh on the React state losing all of the users "session" data

**Review by rpuzon01 — APPROVED** [review]
Great Job

File: api/orders.js (hunk 4 above)
Comment (yeonjupark95) [reply api/orders.js:179]: corrected

**Pull request merged**